### DISCUSSION TITLE
Protected files

### DISCUSSION THREAD

**Discussion by theMackabu** [pocketbase/pocketbase#254] [Q&A] [answered]
Is it possible to protect a file via user profile, for example, only allow access to a file if the user has a specific boolean toggled, or a one time token generated server side?

(Instead of security by obscurity)

**Comment by ganigeorgiev** [comment, marked as answer]
Currently there isn't an easy way to do that. This will be handled as part of https://github.com/pocketbase/pocketbase/issues/215.

For now, if you want to implement it in your application:
1. Create the custom action that will generate this short-lived/one-time token. You could use the `security` subpackage to generate your token, eg:

```go
e.Router.AddRoute(echo.Route{
	Method: http.MethodGet,
	Path:   "/api/users/file-token",
	Handler: func(c echo.Context) error {
		user, _ := c.Get(apis.ContextUserKey).(*models.User)

		token, err := security.NewToken(
			jwt.MapClaims{"id": user.Id, "type": "user"},
			(user.TokenKey + "MY_CUSTOM_TOKEN_KEY"),
			60, // valid for max 60 seconds
		)
		if err != nil {
			return rest.NewBadRequestError("Failed to create token.", err)
		}

		return c.JSON(200, map[string]string{
			"token": token,
		})
	},
	Middlewares: []echo.MiddlewareFunc{
		apis.RequireUserAuth(),
	},
})
```

2. Bind to the OnFileDownloadRequest hook and check the authorized user from there:

```go
app.OnFileDownloadRequest().Add(func(e *core.FileDownloadEvent) error {
    token := e.HttpContext.QueryParam("token")

    user, err := app.Dao().FindUserByToken(token, "MY_CUSTOM_TOKEN_KEY")
    if err != nil || !user.Profiile.GetBoolDataValue("myToggleField") {
        return rest.NewForbiddenError("Forbidden!", err)
    }

    return nil
})
```

**Reply by theMackabu** [reply]
Sounds great, I'll try something akin to the code you provided and also wait for #215

**Reply by ganigeorgiev** [reply]
@theMackabu I've updated the example to include the action to generate your token at `/api/users/file-token`.

**Reply by theMackabu** [reply]
Oh wow tysm
